{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import numpy as np"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# load data"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## load panel df (mapped to CDS already)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "panel_f = '/Users/haochen/Desktop/Tapestri_analysis/panel_analysis/panel3359/CDS/panel3559_analysis.gene_CDS_cytoband.formatted.csv'\n",
    "panel_df = pd.read_csv(panel_f, header = 0)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "panel_df['CDS-exon_unique_id'] = panel_df['CDS-UCSC_ID'] + '-' + panel_df['CDS-UCSC_index']"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## load amplicon sequences"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "amplicon_seq_f = '/Users/haochen/Desktop/Tapestri_analysis/panel_analysis/panel3359/CDS/amplicon_sequences.bed'\n",
    "amplicon_seq_df = pd.read_csv(amplicon_seq_f, header = None, sep = '\\t', names = ['AMPLICON-chr', 'AMPLICON-start', 'AMPLICON-end', 'AMPLICON-ID', 'AMPLICON-seq'])\n",
    "# amplicon_seq_df['AMPLICON-chr'] = amplicon_seq_df[0].str.split(':').str[0]\n",
    "# amplicon_seq_df['AMPLICON-start'] = amplicon_seq_df[0].str.split(':').str[1].str.split('-').str[0]\n",
    "# amplicon_seq_df['AMPLICON-end'] = amplicon_seq_df[0].str.split(':').str[1].str.split('-').str[1]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## get ALL CDS in hg19 genome"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "ALL_CANONICAL_TRANSCRIPTS = '/Users/haochen/Desktop/Tapestri_analysis/resources/gene_annotations/UCSC-hg19-CDS.gene_name_added.sorted.bed'\n",
    "\n",
    "ALL_transcripts_df = pd.read_csv(ALL_CANONICAL_TRANSCRIPTS, header = None, sep = '\\t', names = ['CDS-chr', 'CDS-start', 'CDS-end', 'CDS-Ensembl_ID', 'CDS-UCSC_ID', 'CDS-UCSC_index', 'CDS-strand', 'CDS-gene_name'])"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Subset CDSs to PANEL_GENES"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "canonical_contigs = [('chr' + str(i)) for i in (list(range(1, 23)) + ['X', 'Y'])]\n",
    "\n",
    "ALL_transcripts_df = ALL_transcripts_df[\n",
    "    ALL_transcripts_df['CDS-gene_name'].isin(panel_df['CDS-gene_name'].unique().tolist()) & \n",
    "    ALL_transcripts_df['CDS-chr'].isin(canonical_contigs)\n",
    "    ]\n",
    "ALL_transcripts_df['CDS-exon_unique_id'] = ALL_transcripts_df['CDS-UCSC_ID'] + '-' + ALL_transcripts_df['CDS-UCSC_index'].astype(str)\n",
    "\n",
    "ALL_transcripts_df.to_csv(\n",
    "    '/Users/haochen/Desktop/Tapestri_analysis/resources/gene_annotations/UCSC-hg19-CDS.gene_name_added.sorted.panel_genes.csv', header=True, index = False\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 91,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>CDS-chr</th>\n",
       "      <th>CDS-start</th>\n",
       "      <th>CDS-end</th>\n",
       "      <th>CDS-Ensembl_ID</th>\n",
       "      <th>CDS-UCSC_ID</th>\n",
       "      <th>CDS-UCSC_index</th>\n",
       "      <th>CDS-strand</th>\n",
       "      <th>CDS-gene_name</th>\n",
       "      <th>CDS-exon_unique_id</th>\n",
       "      <th>CDS-ORF-frame</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr12</td>\n",
       "      <td>120554402</td>\n",
       "      <td>120554454</td>\n",
       "      <td>ENST00000229340</td>\n",
       "      <td>uc001txm.2</td>\n",
       "      <td>5</td>\n",
       "      <td>-</td>\n",
       "      <td>RAB35</td>\n",
       "      <td>ENST00000229340.5</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr12</td>\n",
       "      <td>120546220</td>\n",
       "      <td>120546271</td>\n",
       "      <td>ENST00000229340</td>\n",
       "      <td>uc001txm.2</td>\n",
       "      <td>4</td>\n",
       "      <td>-</td>\n",
       "      <td>RAB35</td>\n",
       "      <td>ENST00000229340.4</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr12</td>\n",
       "      <td>120541629</td>\n",
       "      <td>120541753</td>\n",
       "      <td>ENST00000229340</td>\n",
       "      <td>uc001txm.2</td>\n",
       "      <td>3</td>\n",
       "      <td>-</td>\n",
       "      <td>RAB35</td>\n",
       "      <td>ENST00000229340.3</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr12</td>\n",
       "      <td>120536833</td>\n",
       "      <td>120536958</td>\n",
       "      <td>ENST00000229340</td>\n",
       "      <td>uc001txm.2</td>\n",
       "      <td>2</td>\n",
       "      <td>-</td>\n",
       "      <td>RAB35</td>\n",
       "      <td>ENST00000229340.2</td>\n",
       "      <td>2.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr12</td>\n",
       "      <td>120536614</td>\n",
       "      <td>120536739</td>\n",
       "      <td>ENST00000229340</td>\n",
       "      <td>uc001txm.2</td>\n",
       "      <td>1</td>\n",
       "      <td>-</td>\n",
       "      <td>RAB35</td>\n",
       "      <td>ENST00000229340.1</td>\n",
       "      <td>2.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>chr12</td>\n",
       "      <td>120535048</td>\n",
       "      <td>120535177</td>\n",
       "      <td>ENST00000229340</td>\n",
       "      <td>uc001txm.2</td>\n",
       "      <td>0</td>\n",
       "      <td>-</td>\n",
       "      <td>RAB35</td>\n",
       "      <td>ENST00000229340.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>chr12</td>\n",
       "      <td>120534218</td>\n",
       "      <td>120535049</td>\n",
       "      <td>ENST00000229340</td>\n",
       "      <td>uc009zww.2</td>\n",
       "      <td>0</td>\n",
       "      <td>-</td>\n",
       "      <td>RAB35</td>\n",
       "      <td>ENST00000229340.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  CDS-chr  CDS-start    CDS-end   CDS-Ensembl_ID CDS-UCSC_ID  CDS-UCSC_index  \\\n",
       "0   chr12  120554402  120554454  ENST00000229340  uc001txm.2               5   \n",
       "1   chr12  120546220  120546271  ENST00000229340  uc001txm.2               4   \n",
       "2   chr12  120541629  120541753  ENST00000229340  uc001txm.2               3   \n",
       "3   chr12  120536833  120536958  ENST00000229340  uc001txm.2               2   \n",
       "4   chr12  120536614  120536739  ENST00000229340  uc001txm.2               1   \n",
       "5   chr12  120535048  120535177  ENST00000229340  uc001txm.2               0   \n",
       "6   chr12  120534218  120535049  ENST00000229340  uc009zww.2               0   \n",
       "\n",
       "  CDS-strand CDS-gene_name CDS-exon_unique_id  CDS-ORF-frame  \n",
       "0          -         RAB35  ENST00000229340.5            1.0  \n",
       "1          -         RAB35  ENST00000229340.4            0.0  \n",
       "2          -         RAB35  ENST00000229340.3            1.0  \n",
       "3          -         RAB35  ENST00000229340.2            2.0  \n",
       "4          -         RAB35  ENST00000229340.1            2.0  \n",
       "5          -         RAB35  ENST00000229340.0            0.0  \n",
       "6          -         RAB35  ENST00000229340.0            0.0  "
      ]
     },
     "execution_count": 91,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "gene_i = 'RAB35'\n",
    "\n",
    "gene_df = ALL_transcripts_df[ALL_transcripts_df['CDS-gene_name'] == gene_i]\n",
    "#gene_df.reset_index(drop = True, inplace=True)\n",
    "strand = gene_df['CDS-strand'].iloc[0]\n",
    "if strand == '+':\n",
    "    gene_df = gene_df.sort_values(by = ['CDS-start']).reset_index(drop = True)\n",
    "else: # strand == '-':\n",
    "    gene_df = gene_df.sort_values(by = ['CDS-end'], ascending = False).reset_index(drop = True)\n",
    "gene_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>CDS-chr</th>\n",
       "      <th>CDS-start</th>\n",
       "      <th>CDS-end</th>\n",
       "      <th>CDS-Ensembl_ID</th>\n",
       "      <th>CDS-UCSC_ID</th>\n",
       "      <th>CDS-UCSC_index</th>\n",
       "      <th>CDS-strand</th>\n",
       "      <th>CDS-gene_name</th>\n",
       "      <th>CDS-exon_unique_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>887</th>\n",
       "      <td>chr1</td>\n",
       "      <td>6246731</td>\n",
       "      <td>6246876</td>\n",
       "      <td>ENST00000234875</td>\n",
       "      <td>uc001amd.3</td>\n",
       "      <td>0</td>\n",
       "      <td>-</td>\n",
       "      <td>RPL22</td>\n",
       "      <td>uc001amd.3-0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>888</th>\n",
       "      <td>chr1</td>\n",
       "      <td>6252989</td>\n",
       "      <td>6253114</td>\n",
       "      <td>ENST00000234875</td>\n",
       "      <td>uc001amd.3</td>\n",
       "      <td>1</td>\n",
       "      <td>-</td>\n",
       "      <td>RPL22</td>\n",
       "      <td>uc001amd.3-1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>889</th>\n",
       "      <td>chr1</td>\n",
       "      <td>6257711</td>\n",
       "      <td>6257816</td>\n",
       "      <td>ENST00000234875</td>\n",
       "      <td>uc001amd.3</td>\n",
       "      <td>2</td>\n",
       "      <td>-</td>\n",
       "      <td>RPL22</td>\n",
       "      <td>uc001amd.3-2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>890</th>\n",
       "      <td>chr1</td>\n",
       "      <td>6259621</td>\n",
       "      <td>6259633</td>\n",
       "      <td>ENST00000234875</td>\n",
       "      <td>uc001amd.3</td>\n",
       "      <td>3</td>\n",
       "      <td>-</td>\n",
       "      <td>RPL22</td>\n",
       "      <td>uc001amd.3-3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1451</th>\n",
       "      <td>chr1</td>\n",
       "      <td>11167541</td>\n",
       "      <td>11167557</td>\n",
       "      <td>ENST00000361445</td>\n",
       "      <td>uc001asd.3</td>\n",
       "      <td>0</td>\n",
       "      <td>-</td>\n",
       "      <td>MTOR</td>\n",
       "      <td>uc001asd.3-0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>200455</th>\n",
       "      <td>chrX</td>\n",
       "      <td>154319058</td>\n",
       "      <td>154319114</td>\n",
       "      <td>ENST00000369462</td>\n",
       "      <td>uc004fna.3</td>\n",
       "      <td>6</td>\n",
       "      <td>+</td>\n",
       "      <td>BRCC3</td>\n",
       "      <td>uc004fna.3-6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>200456</th>\n",
       "      <td>chrX</td>\n",
       "      <td>154327589</td>\n",
       "      <td>154327664</td>\n",
       "      <td>ENST00000369462</td>\n",
       "      <td>uc004fna.3</td>\n",
       "      <td>7</td>\n",
       "      <td>+</td>\n",
       "      <td>BRCC3</td>\n",
       "      <td>uc004fna.3-7</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>200457</th>\n",
       "      <td>chrX</td>\n",
       "      <td>154344331</td>\n",
       "      <td>154344463</td>\n",
       "      <td>ENST00000369462</td>\n",
       "      <td>uc004fna.3</td>\n",
       "      <td>8</td>\n",
       "      <td>+</td>\n",
       "      <td>BRCC3</td>\n",
       "      <td>uc004fna.3-8</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>200458</th>\n",
       "      <td>chrX</td>\n",
       "      <td>154344985</td>\n",
       "      <td>154345029</td>\n",
       "      <td>ENST00000369462</td>\n",
       "      <td>uc004fna.3</td>\n",
       "      <td>9</td>\n",
       "      <td>+</td>\n",
       "      <td>BRCC3</td>\n",
       "      <td>uc004fna.3-9</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>200459</th>\n",
       "      <td>chrX</td>\n",
       "      <td>154348273</td>\n",
       "      <td>154348425</td>\n",
       "      <td>ENST00000369462</td>\n",
       "      <td>uc004fna.3</td>\n",
       "      <td>10</td>\n",
       "      <td>+</td>\n",
       "      <td>BRCC3</td>\n",
       "      <td>uc004fna.3-10</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5155 rows × 9 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "       CDS-chr  CDS-start    CDS-end   CDS-Ensembl_ID CDS-UCSC_ID  \\\n",
       "887       chr1    6246731    6246876  ENST00000234875  uc001amd.3   \n",
       "888       chr1    6252989    6253114  ENST00000234875  uc001amd.3   \n",
       "889       chr1    6257711    6257816  ENST00000234875  uc001amd.3   \n",
       "890       chr1    6259621    6259633  ENST00000234875  uc001amd.3   \n",
       "1451      chr1   11167541   11167557  ENST00000361445  uc001asd.3   \n",
       "...        ...        ...        ...              ...         ...   \n",
       "200455    chrX  154319058  154319114  ENST00000369462  uc004fna.3   \n",
       "200456    chrX  154327589  154327664  ENST00000369462  uc004fna.3   \n",
       "200457    chrX  154344331  154344463  ENST00000369462  uc004fna.3   \n",
       "200458    chrX  154344985  154345029  ENST00000369462  uc004fna.3   \n",
       "200459    chrX  154348273  154348425  ENST00000369462  uc004fna.3   \n",
       "\n",
       "        CDS-UCSC_index CDS-strand CDS-gene_name CDS-exon_unique_id  \n",
       "887                  0          -         RPL22       uc001amd.3-0  \n",
       "888                  1          -         RPL22       uc001amd.3-1  \n",
       "889                  2          -         RPL22       uc001amd.3-2  \n",
       "890                  3          -         RPL22       uc001amd.3-3  \n",
       "1451                 0          -          MTOR       uc001asd.3-0  \n",
       "...                ...        ...           ...                ...  \n",
       "200455               6          +         BRCC3       uc004fna.3-6  \n",
       "200456               7          +         BRCC3       uc004fna.3-7  \n",
       "200457               8          +         BRCC3       uc004fna.3-8  \n",
       "200458               9          +         BRCC3       uc004fna.3-9  \n",
       "200459              10          +         BRCC3      uc004fna.3-10  \n",
       "\n",
       "[5155 rows x 9 columns]"
      ]
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ALL_transcripts_df"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## fetch ORF information"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 43,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[WARNING] for transcript uc001bfr.3, (gene: ['CDC42']; strand: +), exon 0 is missing. Moving to the next available exon 1...\n",
      "[WARNING] for transcript uc009wgu.3 (gene: ['NRAS']; strand: -), exon 3 is missing. Moving to the next available exon 5...\n",
      "[WARNING] for transcript uc001gcr.1, (gene: ['NUF2']; strand: +), exon 0 is missing. Moving to the next available exon 1...\n",
      "[WARNING] for transcript uc001gpr.3, (gene: ['DHX9']; strand: +), exon 0 is missing. Moving to the next available exon 1...\n",
      "[WARNING] for transcript uc001hdz.2, (gene: ['IKBKE']; strand: +), exon 0 is missing. Moving to the next available exon 2...\n",
      "[WARNING] for transcript uc001kdy.3, (gene: ['BMPR1A']; strand: +), exon 0 is missing. Moving to the next available exon 2...\n",
      "[WARNING] for transcript uc001nmc.4, (gene: ['CTNND1']; strand: +), exon 0 is missing. Moving to the next available exon 2...\n",
      "[WARNING] for transcript uc001pgx.3, (gene: ['BIRC3']; strand: +), exon 0 is missing. Moving to the next available exon 1...\n",
      "[WARNING] for transcript uc001pkb.1, (gene: ['ATM']; strand: +), exon 0 is missing. Moving to the next available exon 1...\n",
      "[WARNING] for transcript uc001rbw.3, (gene: ['ATF7IP']; strand: +), exon 0 is missing. Moving to the next available exon 1...\n",
      "[WARNING] for transcript uc001rdt.3, (gene: ['PIK3C2G']; strand: +), exon 0 is missing. Moving to the next available exon 1...\n",
      "[WARNING] for transcript uc001rgp.1 (gene: ['KRAS']; strand: -), exon 3 is missing. Moving to the next available exon 4...\n",
      "[WARNING] for transcript uc001uub.1, (gene: ['BRCA2']; strand: +), exon 0 is missing. Moving to the next available exon 1...\n",
      "[WARNING] for transcript uc001wqk.2, (gene: ['G2E3']; strand: +), exon 0 is missing. Moving to the next available exon 1...\n",
      "[WARNING] for transcript uc001ymd.2, (gene: ['TRAF3']; strand: +), exon 0 is missing. Moving to the next available exon 1...\n",
      "[WARNING] for transcript uc002aft.4, (gene: ['RNF111']; strand: +), exon 0 is missing. Moving to the next available exon 1...\n",
      "[WARNING] for transcript uc010bnp.1, (gene: ['FANCI']; strand: +), exon 0 is missing. Moving to the next available exon 1...\n",
      "[WARNING] for transcript uc002dwf.3, (gene: ['PPP4C']; strand: +), exon 0 is missing. Moving to the next available exon 1...\n",
      "[WARNING] for transcript uc002eka.3, (gene: ['NUP93']; strand: +), exon 0 is missing. Moving to the next available exon 1...\n",
      "[WARNING] for transcript uc002etl.3, (gene: ['CTCF']; strand: +), exon 0 is missing. Moving to the next available exon 2...\n",
      "[WARNING] for transcript uc002huk.2, (gene: ['RARA']; strand: +), exon 0 is missing. Moving to the next available exon 1...\n",
      "[WARNING] for transcript uc010dbp.2, (gene: ['EME1']; strand: +), exon 0 is missing. Moving to the next available exon 1...\n",
      "[WARNING] for transcript uc002ktt.2, (gene: ['GATA6']; strand: +), exon 0 is missing. Moving to the next available exon 1...\n",
      "[WARNING] for transcript uc010xdp.2, (gene: ['SMAD4']; strand: +), exon 0 is missing. Moving to the next available exon 1...\n",
      "[WARNING] for transcript uc010dxo.3, (gene: ['SMARCA4']; strand: +), exon 0 is missing. Moving to the next available exon 1...\n",
      "[WARNING] for transcript uc002psc.5, (gene: ['POLD1']; strand: +), exon 0 is missing. Moving to the next available exon 1...\n",
      "[WARNING] for transcript uc002rci.3, (gene: ['MYCN']; strand: +), exon 0 is missing. Moving to the next available exon 1...\n",
      "[WARNING] for transcript uc010ysf.2, (gene: ['CTNNA2']; strand: +), exon 0 is missing. Moving to the next available exon 1...\n",
      "[WARNING] for transcript uc002wyc.3, (gene: ['DNMT3B']; strand: +), exon 0 is missing. Moving to the next available exon 1...\n",
      "[WARNING] for transcript uc002xgy.3, (gene: ['SRC']; strand: +), exon 0 is missing. Moving to the next available exon 3...\n",
      "[WARNING] for transcript uc002xtk.3, (gene: ['NCOA3']; strand: +), exon 0 is missing. Moving to the next available exon 2...\n",
      "[WARNING] for transcript uc011abd.2, (gene: ['RTEL1']; strand: +), exon 0 is missing. Moving to the next available exon 1...\n",
      "[WARNING] for transcript uc003bao.1, (gene: ['XRCC6']; strand: +), exon 0 is missing. Moving to the next available exon 1...\n",
      "[WARNING] for transcript uc003buw.3, (gene: ['FANCD2']; strand: +), exon 0 is missing. Moving to the next available exon 1...\n",
      "[WARNING] for transcript uc003ckr.2, (gene: ['CTNNB1']; strand: +), exon 0 is missing. Moving to the next available exon 1...\n",
      "[WARNING] for transcript uc003fjk.3, (gene: ['PIK3CA']; strand: +), exon 0 is missing. Moving to the next available exon 1...\n",
      "[WARNING] for transcript uc003gdu.2, (gene: ['FGFR3']; strand: +), exon 0 is missing. Moving to the next available exon 1...\n",
      "[WARNING] for transcript uc003gdz.4, (gene: ['WHSC1']; strand: +), exon 0 is missing. Moving to the next available exon 1...\n",
      "[WARNING] for transcript uc003han.4, (gene: ['PDGFRA']; strand: +), exon 0 is missing. Moving to the next available exon 1...\n",
      "[WARNING] for transcript uc003hxk.3, (gene: ['TET2']; strand: +), exon 0 is missing. Moving to the next available exon 2...\n",
      "[WARNING] for transcript uc003hyt.2 (gene: ['LEF1']; strand: -), exon 10 is missing. Moving to the next available exon 11...\n",
      "[WARNING] for transcript uc003jdx.1, (gene: ['PAPD7']; strand: +), exon 0 is missing. Moving to the next available exon 1...\n",
      "[WARNING] for transcript uc003jkl.4, (gene: ['NIPBL']; strand: +), exon 0 is missing. Moving to the next available exon 1...\n",
      "[WARNING] for transcript uc003kpy.4, (gene: ['APC']; strand: +), exon 0 is missing. Moving to the next available exon 1...\n",
      "[WARNING] for transcript uc003ldh.3, (gene: ['CTNNA1']; strand: +), exon 0 is missing. Moving to the next available exon 1...\n",
      "[WARNING] for transcript uc003msz.4, (gene: ['IRF4']; strand: +), exon 0 is missing. Moving to the next available exon 1...\n",
      "[WARNING] for transcript uc003mxb.3, (gene: ['RREB1']; strand: +), exon 0 is missing. Moving to the next available exon 3...\n",
      "[WARNING] for transcript uc003nwu.2, (gene: ['MSH5-SAPCD1']; strand: +), exon 0 is missing. Moving to the next available exon 1...\n",
      "[WARNING] for transcript uc003nwz.4, (gene: ['MSH5-SAPCD1']; strand: +), exon 0 is missing. Moving to the next available exon 9...\n",
      "[WARNING] for transcript uc003pjc.3, (gene: ['TTK']; strand: +), exon 0 is missing. Moving to the next available exon 1...\n",
      "[WARNING] for transcript uc010ljc.3, (gene: ['KMT2E']; strand: +), exon 0 is missing. Moving to the next available exon 1...\n",
      "[WARNING] for transcript uc010lkh.3, (gene: ['MET']; strand: +), exon 0 is missing. Moving to the next available exon 1...\n",
      "[WARNING] for transcript uc003xio.4, (gene: ['WRN']; strand: +), exon 0 is missing. Moving to the next available exon 1...\n",
      "[WARNING] for transcript uc003xsk.4, (gene: ['LYN']; strand: +), exon 0 is missing. Moving to the next available exon 1...\n",
      "[WARNING] for transcript uc003zpl.3 (gene: ['CDKN2A']; strand: -), exon 1 is missing. Moving to the next available exon 2...\n",
      "[WARNING] for transcript uc004bvm.4, (gene: ['SPTAN1']; strand: +), exon 0 is missing. Moving to the next available exon 1...\n",
      "[WARNING] for transcript uc004dic.2, (gene: ['ARAF']; strand: +), exon 0 is missing. Moving to the next available exon 1...\n",
      "[WARNING] for transcript uc004eun.4 (gene: ['SMARCA1']; strand: -), exon 23 is missing. Moving to the next available exon 24...\n"
     ]
    }
   ],
   "source": [
    "# transcript_bp_df = pd.DataFrame(index = ALL_transcripts_df['CDS-exon_unique_id'].unique(), columns = ['CDS-length'])\n",
    "for transcript_i in ALL_transcripts_df['CDS-UCSC_ID'].unique():\n",
    "    # print(transcript_i)\n",
    "    transcript_df = ALL_transcripts_df[ALL_transcripts_df['CDS-UCSC_ID'] == transcript_i]\n",
    "    gene_name = transcript_df['CDS-gene_name'].unique()\n",
    "    # # ----- multiple transcripts (DEPRECATED SINCE THIS ONLY APPLIES WHEN ITERRATING THROUGH UNIQUE GENES) -----\n",
    "    # if len( transcript_df['CDS-UCSC_ID'].unique() ) > 1:\n",
    "    #     print(f'[WARNING] {transcript_i} has multiple transcripts; taking the longest one')\n",
    "    #     longest_transcript = transcript_df['CDS-UCSC_ID'].value_counts().idxmax()\n",
    "    #     print(f'--- longest transcript: {longest_transcript}')\n",
    "    #     transcript_df = transcript_df[transcript_df['CDS-UCSC_ID'] == longest_transcript]\n",
    "\n",
    "    # ----- strand -----\n",
    "    strand = transcript_df['CDS-strand'].iloc[0]\n",
    "    if strand == '+':\n",
    "        transcript_df = transcript_df.sort_values(by = ['CDS-start']).reset_index(drop = True)\n",
    "    else: # strand == '-':\n",
    "        transcript_df = transcript_df.sort_values(by = ['CDS-end'], ascending = False).reset_index(drop = True)\n",
    "        num_exons = len(transcript_df)\n",
    "        \n",
    "    length = 0\n",
    "    cds_idx = 0\n",
    "    redundant_bp_sum = 0\n",
    "    for _, row in transcript_df.iterrows():\n",
    "        if strand == '+' and not (cds_idx == row['CDS-UCSC_index']):\n",
    "            # print(f'transcript is {transcript_i}')\n",
    "            # print(f'cds_idx is {cds_idx}')\n",
    "            # print(f'num_exons is {num_exons}')\n",
    "            # print(f'cds idx is {row[\"CDS-UCSC_index\"]}')\n",
    "            # raise ValueError\n",
    "            print(f'[WARNING] for transcript {transcript_i}, (gene: {gene_name}; strand: {strand}), exon {cds_idx} is missing. Moving to the next available exon {row[\"CDS-UCSC_index\"]}...')\n",
    "            cds_idx = row['CDS-UCSC_index']\n",
    "        if strand == '-' and not ( cds_idx == (num_exons - 1 - row['CDS-UCSC_index']) ):\n",
    "            # print(f'transcript is {transcript_i}')\n",
    "            # print(f'cds_idx is {cds_idx}')\n",
    "            # print(f'num_exons is {num_exons}')\n",
    "            # print(f'cds idx is {row[\"CDS-UCSC_index\"]}')\n",
    "            # raise ValueError\n",
    "            print(f'[WARNING] for transcript {transcript_i} (gene: {gene_name}; strand: {strand}), exon {(num_exons - 1)} is missing. Moving to the next available exon {row[\"CDS-UCSC_index\"]}...')\n",
    "            cds_idx = num_exons - 1 - row['CDS-UCSC_index']\n",
    "        unique_id = row['CDS-exon_unique_id']\n",
    "        \n",
    "        # print(f'length: {length}')\n",
    "        cds_idx += 1\n",
    "        # --- count redundant bp at the beginning ---\n",
    "        start_r_bp = (length - redundant_bp_sum) % 3\n",
    "        redundant_bp_sum += start_r_bp\n",
    "        # print(f'start redundant bp: {start_r_bp}')\n",
    "        # --- count redundant bp at the end ---\n",
    "        length += row['CDS-end'] - row['CDS-start']\n",
    "        end_r_bp = (length - redundant_bp_sum) % 3\n",
    "        redundant_bp_sum += end_r_bp\n",
    "        # print(f'end redundant bp: {end_r_bp}')\n",
    "\n",
    "    \n",
    "        ALL_transcripts_df.loc[\n",
    "            (ALL_transcripts_df['CDS-exon_unique_id'] == unique_id),\n",
    "            ['start_r_bp', 'end_r_bp', 'redundant_bp_sum']\n",
    "        ] = [start_r_bp, end_r_bp, redundant_bp_sum]\n",
    "        \n",
    "    \n",
    "    # redundant_bp_sum = ALL_transcripts_df[ALL_transcripts_df['CDS-UCSC_ID'] == transcript_i]['redundant_bp_sum'].sum()\n",
    "    if not redundant_bp_sum % 3 == 0:\n",
    "        print(f'for transcript {transcript_i}, total length is {length}')\n",
    "        raise ValueError(f'ORF frame sum for transcript {transcript_i} is not divisible by 3.')\n",
    "\n",
    "    # transcript_bp_df.loc[transcript_i, 'CDS-length'] = length"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 44,
   "metadata": {},
   "outputs": [],
   "source": [
    "ALL_transcripts_df.to_csv(\n",
    "    '/Users/haochen/Desktop/Tapestri_analysis/resources/gene_annotations/UCSC-hg19-CDS.gene_name_added.sorted.panel_genes.orf_added.csv', header=True, index = False\n",
    ")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## add ORF info to panel_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 45,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>INSERT-chr</th>\n",
       "      <th>INSERT-start</th>\n",
       "      <th>INSERT-end</th>\n",
       "      <th>AMPLICON-ID</th>\n",
       "      <th>CDS-chr</th>\n",
       "      <th>CDS-start</th>\n",
       "      <th>CDS-end</th>\n",
       "      <th>CDS-Ensembl_ID</th>\n",
       "      <th>CDS-UCSC_ID</th>\n",
       "      <th>CDS-UCSC_index</th>\n",
       "      <th>CDS-UCSC_strand</th>\n",
       "      <th>CDS-gene_name</th>\n",
       "      <th>bp_overlap</th>\n",
       "      <th>overlap_start</th>\n",
       "      <th>overlap_end</th>\n",
       "      <th>CDS-exon_unique_id</th>\n",
       "      <th>CDS-ORF-frame</th>\n",
       "      <th>start_r_bp</th>\n",
       "      <th>end_r_bp</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>6252892</td>\n",
       "      <td>6253107</td>\n",
       "      <td>AMPL250217</td>\n",
       "      <td>chr1</td>\n",
       "      <td>6252989</td>\n",
       "      <td>6253114</td>\n",
       "      <td>ENST00000234875</td>\n",
       "      <td>uc001amd.3</td>\n",
       "      <td>1</td>\n",
       "      <td>-</td>\n",
       "      <td>RPL22</td>\n",
       "      <td>118</td>\n",
       "      <td>6252989</td>\n",
       "      <td>6253107</td>\n",
       "      <td>uc001amd.3-1</td>\n",
       "      <td>2.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>2.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1</td>\n",
       "      <td>6257606</td>\n",
       "      <td>6257826</td>\n",
       "      <td>AMPL257565</td>\n",
       "      <td>chr1</td>\n",
       "      <td>6257711</td>\n",
       "      <td>6257816</td>\n",
       "      <td>ENST00000234875</td>\n",
       "      <td>uc001amd.3</td>\n",
       "      <td>2</td>\n",
       "      <td>-</td>\n",
       "      <td>RPL22</td>\n",
       "      <td>105</td>\n",
       "      <td>6257711</td>\n",
       "      <td>6257816</td>\n",
       "      <td>uc001amd.3-2</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr1</td>\n",
       "      <td>11204888</td>\n",
       "      <td>11205104</td>\n",
       "      <td>AMPL257566</td>\n",
       "      <td>chr1</td>\n",
       "      <td>11205024</td>\n",
       "      <td>11205102</td>\n",
       "      <td>ENST00000361445</td>\n",
       "      <td>uc001asd.3</td>\n",
       "      <td>25</td>\n",
       "      <td>-</td>\n",
       "      <td>MTOR</td>\n",
       "      <td>78</td>\n",
       "      <td>11205024</td>\n",
       "      <td>11205102</td>\n",
       "      <td>uc001asd.3-25</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr1</td>\n",
       "      <td>11288734</td>\n",
       "      <td>11288957</td>\n",
       "      <td>AMPL253201</td>\n",
       "      <td>chr1</td>\n",
       "      <td>11288724</td>\n",
       "      <td>11288975</td>\n",
       "      <td>ENST00000361445</td>\n",
       "      <td>uc001asd.3</td>\n",
       "      <td>39</td>\n",
       "      <td>-</td>\n",
       "      <td>MTOR</td>\n",
       "      <td>223</td>\n",
       "      <td>11288734</td>\n",
       "      <td>11288957</td>\n",
       "      <td>uc001asd.3-39</td>\n",
       "      <td>2.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>2.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr1</td>\n",
       "      <td>11297897</td>\n",
       "      <td>11298107</td>\n",
       "      <td>AMPL257567</td>\n",
       "      <td>chr1</td>\n",
       "      <td>11297899</td>\n",
       "      <td>11298105</td>\n",
       "      <td>ENST00000361445</td>\n",
       "      <td>uc001asd.3</td>\n",
       "      <td>45</td>\n",
       "      <td>-</td>\n",
       "      <td>MTOR</td>\n",
       "      <td>206</td>\n",
       "      <td>11297899</td>\n",
       "      <td>11298105</td>\n",
       "      <td>uc001asd.3-45</td>\n",
       "      <td>2.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>2.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>630</th>\n",
       "      <td>chrX</td>\n",
       "      <td>128605116</td>\n",
       "      <td>128605306</td>\n",
       "      <td>AMPL257734</td>\n",
       "      <td>chrX</td>\n",
       "      <td>128605180</td>\n",
       "      <td>128605303</td>\n",
       "      <td>ENST00000371122</td>\n",
       "      <td>uc004eun.4</td>\n",
       "      <td>5</td>\n",
       "      <td>-</td>\n",
       "      <td>SMARCA1</td>\n",
       "      <td>123</td>\n",
       "      <td>128605180</td>\n",
       "      <td>128605303</td>\n",
       "      <td>uc004eun.4-5</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>631</th>\n",
       "      <td>chrX</td>\n",
       "      <td>135957311</td>\n",
       "      <td>135957532</td>\n",
       "      <td>AMPL250434</td>\n",
       "      <td>chrX</td>\n",
       "      <td>135957244</td>\n",
       "      <td>135957327</td>\n",
       "      <td>ENST00000320676</td>\n",
       "      <td>uc004fae.2</td>\n",
       "      <td>1</td>\n",
       "      <td>-</td>\n",
       "      <td>RBMX</td>\n",
       "      <td>16</td>\n",
       "      <td>135957311</td>\n",
       "      <td>135957327</td>\n",
       "      <td>uc004fae.2-1</td>\n",
       "      <td>2.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>2.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>632</th>\n",
       "      <td>chrX</td>\n",
       "      <td>135957311</td>\n",
       "      <td>135957532</td>\n",
       "      <td>AMPL250434</td>\n",
       "      <td>chrX</td>\n",
       "      <td>135957416</td>\n",
       "      <td>135957542</td>\n",
       "      <td>ENST00000320676</td>\n",
       "      <td>uc004fae.2</td>\n",
       "      <td>2</td>\n",
       "      <td>-</td>\n",
       "      <td>RBMX</td>\n",
       "      <td>116</td>\n",
       "      <td>135957416</td>\n",
       "      <td>135957532</td>\n",
       "      <td>uc004fae.2-2</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>633</th>\n",
       "      <td>chrX</td>\n",
       "      <td>153581295</td>\n",
       "      <td>153581504</td>\n",
       "      <td>AMPL250435</td>\n",
       "      <td>chrX</td>\n",
       "      <td>153581368</td>\n",
       "      <td>153581572</td>\n",
       "      <td>ENST00000422373</td>\n",
       "      <td>uc004fkk.2</td>\n",
       "      <td>10</td>\n",
       "      <td>-</td>\n",
       "      <td>FLNA</td>\n",
       "      <td>136</td>\n",
       "      <td>153581368</td>\n",
       "      <td>153581504</td>\n",
       "      <td>uc004fkk.2-10</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>634</th>\n",
       "      <td>chrX</td>\n",
       "      <td>154299804</td>\n",
       "      <td>154299956</td>\n",
       "      <td>AMPL251376</td>\n",
       "      <td>chrX</td>\n",
       "      <td>154299802</td>\n",
       "      <td>154299925</td>\n",
       "      <td>ENST00000369462</td>\n",
       "      <td>uc004fna.3</td>\n",
       "      <td>0</td>\n",
       "      <td>+</td>\n",
       "      <td>BRCC3</td>\n",
       "      <td>121</td>\n",
       "      <td>154299804</td>\n",
       "      <td>154299925</td>\n",
       "      <td>uc004fna.3-0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>635 rows × 19 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "    INSERT-chr  INSERT-start  INSERT-end AMPLICON-ID CDS-chr  CDS-start  \\\n",
       "0         chr1       6252892     6253107  AMPL250217    chr1    6252989   \n",
       "1         chr1       6257606     6257826  AMPL257565    chr1    6257711   \n",
       "2         chr1      11204888    11205104  AMPL257566    chr1   11205024   \n",
       "3         chr1      11288734    11288957  AMPL253201    chr1   11288724   \n",
       "4         chr1      11297897    11298107  AMPL257567    chr1   11297899   \n",
       "..         ...           ...         ...         ...     ...        ...   \n",
       "630       chrX     128605116   128605306  AMPL257734    chrX  128605180   \n",
       "631       chrX     135957311   135957532  AMPL250434    chrX  135957244   \n",
       "632       chrX     135957311   135957532  AMPL250434    chrX  135957416   \n",
       "633       chrX     153581295   153581504  AMPL250435    chrX  153581368   \n",
       "634       chrX     154299804   154299956  AMPL251376    chrX  154299802   \n",
       "\n",
       "       CDS-end   CDS-Ensembl_ID CDS-UCSC_ID CDS-UCSC_index CDS-UCSC_strand  \\\n",
       "0      6253114  ENST00000234875  uc001amd.3              1               -   \n",
       "1      6257816  ENST00000234875  uc001amd.3              2               -   \n",
       "2     11205102  ENST00000361445  uc001asd.3             25               -   \n",
       "3     11288975  ENST00000361445  uc001asd.3             39               -   \n",
       "4     11298105  ENST00000361445  uc001asd.3             45               -   \n",
       "..         ...              ...         ...            ...             ...   \n",
       "630  128605303  ENST00000371122  uc004eun.4              5               -   \n",
       "631  135957327  ENST00000320676  uc004fae.2              1               -   \n",
       "632  135957542  ENST00000320676  uc004fae.2              2               -   \n",
       "633  153581572  ENST00000422373  uc004fkk.2             10               -   \n",
       "634  154299925  ENST00000369462  uc004fna.3              0               +   \n",
       "\n",
       "    CDS-gene_name  bp_overlap  overlap_start  overlap_end CDS-exon_unique_id  \\\n",
       "0           RPL22         118        6252989      6253107       uc001amd.3-1   \n",
       "1           RPL22         105        6257711      6257816       uc001amd.3-2   \n",
       "2            MTOR          78       11205024     11205102      uc001asd.3-25   \n",
       "3            MTOR         223       11288734     11288957      uc001asd.3-39   \n",
       "4            MTOR         206       11297899     11298105      uc001asd.3-45   \n",
       "..            ...         ...            ...          ...                ...   \n",
       "630       SMARCA1         123      128605180    128605303       uc004eun.4-5   \n",
       "631          RBMX          16      135957311    135957327       uc004fae.2-1   \n",
       "632          RBMX         116      135957416    135957532       uc004fae.2-2   \n",
       "633          FLNA         136      153581368    153581504      uc004fkk.2-10   \n",
       "634         BRCC3         121      154299804    154299925       uc004fna.3-0   \n",
       "\n",
       "    CDS-ORF-frame  start_r_bp  end_r_bp  \n",
       "0             2.0         0.0       2.0  \n",
       "1             0.0         0.0       0.0  \n",
       "2             0.0         0.0       0.0  \n",
       "3             2.0         0.0       2.0  \n",
       "4             2.0         0.0       2.0  \n",
       "..            ...         ...       ...  \n",
       "630           0.0         0.0       0.0  \n",
       "631           2.0         0.0       2.0  \n",
       "632           0.0         0.0       0.0  \n",
       "633           0.0         0.0       0.0  \n",
       "634           0.0         0.0       0.0  \n",
       "\n",
       "[635 rows x 19 columns]"
      ]
     },
     "execution_count": 45,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# panel_df['CDS-ORF-frame'] = panel_df['CDS-exon_unique_id'].map(\n",
    "#     lambda unique_exon: \n",
    "#     ALL_transcripts_df.loc[\n",
    "#         ALL_transcripts_df['CDS-exon_unique_id'] == unique_exon,\n",
    "#         'CDS-ORF-frame'\n",
    "#         ].values[0]).astype(int)\n",
    "\n",
    "for _, row in panel_df.iterrows():\n",
    "    exon_unique_id = row['CDS-exon_unique_id']\n",
    "    if exon_unique_id == '.-.': # if the amplicon does not cover anything\n",
    "        continue\n",
    "    gene_name = row['CDS-gene_name']\n",
    "    orf_frame_info = ALL_transcripts_df.loc[\n",
    "        ALL_transcripts_df['CDS-exon_unique_id'] == exon_unique_id,\n",
    "        ['start_r_bp', 'end_r_bp']\n",
    "    ].values\n",
    "    if orf_frame_info.shape[0] == 0:\n",
    "        print(f'no orf frame for {gene_name} - {exon_unique_id}')\n",
    "        raise ValueError\n",
    "    if orf_frame_info.shape[0] > 1:\n",
    "        print(f'multiple orf frames for {gene_name} - {exon_unique_id}')\n",
    "        raise ValueError\n",
    "    panel_df.loc[\n",
    "        (panel_df['CDS-exon_unique_id'] == exon_unique_id),\n",
    "        ['start_r_bp', 'end_r_bp']\n",
    "    ] = orf_frame_info\n",
    "panel_df"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# examine the size of coverage for each gene"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "gene_map = {}\n",
    "for _, row in panel_df.iterrows():\n",
    "    gene_name = row['CDS-gene_name']\n",
    "    if not gene_name in gene_map:\n",
    "        gene_map[gene_name] = int(row['bp_overlap'])\n",
    "    else:\n",
    "        gene_map[gene_name] += int(row['bp_overlap'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "pd.DataFrame.from_dict(gene_map, orient = 'index', columns = ['bp_CDS_covered']).sort_values(by = 'bp_CDS_covered', ascending = False).head(40)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# write the `build_table` function\n",
    "\n",
    "Following Martincorena et al., Cell (2017), for each gene's CDS, covered by panel3359, we want to build an 192 by 4 matrix:\n",
    "- 192 unique trinucleotide-context-aware mutations * 4 unique mutation types (synonymous, missense, nonsense, splice)\n",
    "Each cell of this matrix would be the number of possible mutations"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Precalculate a 64x64 matrix with the functional impact of each codon transition (1=Synonymous, 2=Missense, 3=Nonsense)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 372,
   "metadata": {},
   "outputs": [],
   "source": [
    "# from tea.snv import translate\n",
    "\n",
    "# Precalculating a 64x64 matrix with the functional impact of each codon transition (1=Synonymous, 2=Missense, 3=Nonsense)\n",
    "\n",
    "trinuc_list = []\n",
    "for i in ['A', 'C', 'G', 'T']:\n",
    "    for j in ['A', 'C', 'G', 'T']:\n",
    "        for k in ['A', 'C', 'G', 'T']:\n",
    "            trinuc_list.append(i+j+k)\n",
    "\n",
    "IMPACT_MATRIX = pd.DataFrame(np.zeros((64,64)))\n",
    "IMPACT_MATRIX.columns = IMPACT_MATRIX.index = trinuc_list\n",
    "\n",
    "\n",
    "for trinuc_i in IMPACT_MATRIX.index:\n",
    "    for trinuc_j in IMPACT_MATRIX.columns:\n",
    "        from_aa = translate(trinuc_i)\n",
    "        to_aa = translate(trinuc_j)\n",
    "        if from_aa == to_aa:\n",
    "            IMPACT_MATRIX.loc[trinuc_i, trinuc_j] = 1\n",
    "        elif to_aa == '*':\n",
    "            IMPACT_MATRIX.loc[trinuc_i, trinuc_j] = 3\n",
    "        elif to_aa != '*' and from_aa != '*' and to_aa != from_aa:\n",
    "            IMPACT_MATRIX.loc[trinuc_i, trinuc_j] = 2\n",
    "        else: # from _aa == '*'\n",
    "            IMPACT_MATRIX.loc[trinuc_i, trinuc_j] = np.nan\n",
    "\n",
    "# IMPACT_MATRIX = IMPACT_MATRIX.astype(int)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## functions to paste strings, mutate codons"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "from functools import reduce\n",
    "def __reduce_concat(x, sep=\"\"):\n",
    "    return reduce(lambda x, y: str(x) + sep + str(y), x)\n",
    "\n",
    "def paste(*lists, sep=\" \", collapse=None):\n",
    "    result = map(lambda x: __reduce_concat(x, sep=sep), zip(*lists))\n",
    "    if collapse is not None:\n",
    "        return __reduce_concat(result, sep=collapse)\n",
    "    return list(result)\n",
    "\n",
    "def __mutate(codon, pos, nt):\n",
    "    '''\n",
    "    Mutate a codon at a given position to a given nucleotide\n",
    "\n",
    "    Parameters\n",
    "    ----------\n",
    "    codon : str\n",
    "        The codon to mutate\n",
    "    pos : int\n",
    "        The position to mutate (1, 2, or 3)\n",
    "    nt : str\n",
    "        The nucleotide to mutate to\n",
    "\n",
    "    Returns\n",
    "    -------\n",
    "    str\n",
    "    '''\n",
    "    if not len(codon) == 3:\n",
    "        raise ValueError\n",
    "    if not pos in [0, 1, 2]:\n",
    "        raise ValueError\n",
    "    if not nt in ['A', 'C', 'G', 'T']:\n",
    "        raise ValueError\n",
    "    mutated_codon = codon[:pos] + nt + codon[pos+1:]\n",
    "    return mutated_codon"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## initialize trinucleotide list"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "trinuc_list = paste(\n",
    "    np.repeat(nts, 16),\n",
    "    np.repeat(np.tile(nts, 4), 4),\n",
    "    np.tile(nts, 16),\n",
    "    sep = '',\n",
    ")\n",
    "\n",
    "trinuc_ind = pd.Series(range(64), index = trinuc_list) # all the trinucleotide mutation patterns. Should be 4 * 3 * (4^2) = 192 \n",
    "trinuc_subs = np.array(\n",
    "    paste(\n",
    "        np.repeat(nts, 48),\n",
    "        np.tile(np.repeat(nts, 3), 16),\n",
    "        np.tile(np.repeat(nts, 12), 4),\n",
    "        np.repeat('>', 192),\n",
    "        np.repeat(nts, 48),\n",
    "        np.array([nts[nts != j].tolist() for j in np.tile(nts, 16)]).flatten(),\n",
    "        np.tile(np.repeat(nts, 12), 4),\n",
    "        sep = '',\n",
    "    )\n",
    ")\n",
    "trinuc_subsind = pd.Series(range(192), index=trinuc_subs)\n",
    "# sanity check\n",
    "np.unique(np.array(trinuc_subs)).shape"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## create L matrices for each CDS at each trinucleotide context: \n",
    "- number of synonymous\n",
    "- number of missense\n",
    "- number of nonsense\n",
    "- number of splice sites"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## test case: first coding sequence in dndscv refcds_hg19.rda"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 490,
   "metadata": {},
   "outputs": [],
   "source": [
    "# seq = 'CATGTCCA'\n",
    "seq = 'cATGTCCATGCTCGTGGTCTTTCTCTTGCTGTGGGGTGTCACCTGGGGCCCAGTGACAGAAGCAGCCATATTTTATGAGACGCAGCCCAGCCTGTGGGCAGAGTCCGAATCACTGCTGAAACCCTTGGCCAATGTGACGCTGACGTGCCAGGCCCACCTGGAGACTCCAGACTTCCAGCTGTTCAAGAATGGGGTGGCCCAGGAGCCTGTGCACCTTGACTCACCTGCCATCAAGCACCAGTTCCTGCTGACGGGTGACACCCAGGGCCGCTACCGCTGCCGCTCGGGCTTGTCCACAGGATGGACCCAGCTGAGCAAGCTCCTGGAGCTGACAGGGCCAAAGTCCTTGCCTGCTCCCTGGCTCTCGATGGCGCCAGTGTCCTGGATCACCCCCGGCCTGAAAACAACAGCAGTGTGCCGAGGTGTGCTGCGGGGTGTGACTTTTCTGCTGAGGCGGGAGGGCGACCATGAGTTTCTGGAGGTGCCTGAGGCCCAGGAGGATGTGGAGGCCACCTTTCCAGTCCATCAGCCTGGCAACTACAGCTGCAGCTACCGGACCGATGGGGAAGGCGCCCTCTCTGAGCCCAGCGCTACTGTGACCATTGAGGAGCTCGCTGCACCACCACCGCCTGTGCTGATGCACCATGGAGAGTCCTCCCAGGTCCTGCACCCTGGCAACAAGGTGACCCTCACCTGCGTGGCTCCCCTGAGTGGAGTGGACTTCCAGCTACGGCGCGGGGAGAAAGAGCTGCTGGTACCCAGGAGCAGCACCAGCCCAGATCGCATCTTCTTTCACCTGAACGCGGTGGCCCTGGGGGATGGAGGTCACTACACCTGCCGCTACCGGCTGCATGACAACCAAAACGGCTGGTCCGGGGACAGCGCGCCGGTCGAGCTGATTCTGAGCGATGAGACGCTGCCCGCGCCGGAGTTCTCCCCGGAGCCGGAGTCCGGCAGGGCCTTGCGGCTGCGGTGCCTGGCGCCCCTGGAGGGCGCGCGCTTCGCCCTGGTGCGCGAGGACAGGGGCGGGCGCCGCGTGCACCGTTTCCAGAGCCCCGCTGGGACCGAGGCGCTCTTCGAGCTGCACAACATTTCCGTGGCTGACTCCGCCAACTACAGCTGCGTCTACGTGGACCTGAAGCCGCCTTTCGGGGGCTCCGCGCCCAGCGAGCGCTTGGAGCTGCACGTGGACGGACCCCCTCCCAGGCCTCAGCTCCGGGCGACGTGGAGTGGGGCGGTCCTGGCGGGCCGAGATGCCGTCCTGCGCTGCGAGGGACCCATCCCCGACGTCACCTTCGAGCTGCTGCGCGAGGGCGAGACGAAGGCCGTGAAGACGGTCCGCACCCCCGGGGCCGCGGCGAACCTCGAGCTGATCTTCGTGGGGCCCCAGCACGCCGGCAACTACAGGTGCCGCTACCGCTCCTGGGTGCCCCACACCTTCGAATCGGAGCTCAGCGACCCTGTGGAGCTCCTGGTGGCAGAAAGCTGAt'\n",
    "\n",
    "seq = [i for i in seq.upper()]\n",
    "start = 1\n",
    "end = len(seq) - 1\n",
    "seq_1up = np.array( seq[start-1:end-1] ) # 1-bp upstream e.g. ATG\n",
    "seq_mid = np.array( seq[start:end] ) # middle e.g. TGC\n",
    "seq_1down = np.array( seq[start+1:end+1] ) # 1-bp downstream e.g. GCA\n",
    "seq_length = len(seq_mid)\n",
    "\n",
    "##### ----- trinucleotide context ----- #####\n",
    "indices = np.repeat([i for i in range(seq_length)], 3) # times 3 because for each single nucleotide, there are 3 ways to mutate it; e.g. array([0, 0, 0, 1, 1, 1, 2, 2, 2, 3, 3, 3, 4, 4, 4, 5, 5, 5])\n",
    "old_trinuc = np.array(\n",
    "    paste(seq_1up[indices], seq_mid[indices], seq_1down[indices], sep = \"\") \n",
    "    ) # array(['CAT', 'CAT', 'CAT', 'ATG', 'ATG', 'ATG', 'TGT', 'TGT', 'TGT','GTC', 'GTC', 'GTC', 'TCC', 'TCC', 'TCC', 'CCA', 'CCA', 'CCA'])\n",
    "new_bases = np.array(\n",
    "    [nts[nts != x].tolist() for x in seq_mid] \n",
    "    ).flatten() # generate all possible new bases at each position e.g. array(['C', 'G', 'T', 'A', 'C', 'G', 'A', 'C', 'T', 'A', 'C', 'G', 'A', 'G', 'T', 'A', 'G', 'T'])\n",
    "new_trinuc = np.array(\n",
    "    paste(seq_1up[indices], new_bases, seq_1down[indices], sep=\"\") \n",
    "    ) # array(['CCT', 'CGT', 'CTT', 'AAG', 'ACG', 'AGG', 'TAT', 'TCT', 'TTT', 'GAC', 'GCC', 'GGC', 'TAC', 'TGC', 'TTC', 'CAA', 'CGA', 'CTA'])\n",
    "\n",
    "##### ----- codon ----- #####\n",
    "num_codon = int(seq_length / 3)\n",
    "codon_start = np.repeat([i for i in range(0, seq_length, 3)], 9) # 9 here because for each codon, there are 9 possible ways to mutate it\n",
    "# e.g. array([0, 0, 0, 0, 0, 0, 0, 0, 0, 3, 3, 3, 3, 3, 3, 3, 3, 3])\n",
    "\n",
    "old_codon = np.array(\n",
    "    paste(seq_mid[codon_start], seq_mid[codon_start+1], seq_mid[codon_start+2], sep=\"\")\n",
    "    ) # array(['ATG', 'ATG', 'ATG', 'ATG', 'ATG', 'ATG', 'ATG', 'ATG', 'ATG', 'TCC', 'TCC', 'TCC', 'TCC', 'TCC', 'TCC', 'TCC', 'TCC', 'TCC'])\n",
    "pos_in_codon_to_mutate = np.tile(np.repeat([i for i in range(3)], 3), num_codon) # e.g. array([0, 0, 0, 1, 1, 1, 2, 2, 2, 0, 0, 0, 1, 1, 1, 2, 2, 2])\n",
    "\n",
    "new_codon = np.array([__mutate(old_codon[x], pos_in_codon_to_mutate[x], new_bases[x]) for x in range(len(pos_in_codon_to_mutate))]) # e.g. array(['CTG', 'GTG', 'TTG', 'AAG', 'ACG', 'AGG', 'ATA', 'ATC', 'ATT', 'ACC', 'CCC', 'GCC', 'TAC', 'TGC', 'TTC', 'TCA', 'TCG', 'TCT']\n",
    "\n",
    "imp = np.array([ \n",
    "    IMPACT_MATRIX.loc[old_codon[x], new_codon[x]] \n",
    "    for x in range(len(old_codon)) \n",
    "    ]) # array([2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 1, 1, 1])\n",
    "\n",
    "##### ----- create L matrix ----- #####\n",
    "trinuc_subs_oi = pd.DataFrame(\n",
    "    data = imp,\n",
    "    index = paste(old_trinuc, new_trinuc, sep=\">\"),\n",
    "    columns = ['IMPACT']\n",
    "    )\n",
    "\n",
    "L = pd.DataFrame(index = trinuc_subs, columns = ['Synonymous', 'Missense', 'Nonsense'])\n",
    "# Synonymous\n",
    "val_counts = trinuc_subs_oi[imp==1].index.value_counts()\n",
    "L.loc[ val_counts.index, 'Synonymous' ] = val_counts.values\n",
    "\n",
    "# Missense\n",
    "val_counts = trinuc_subs_oi[imp==2].index.value_counts()\n",
    "L.loc[ val_counts.index, 'Missense' ] = val_counts.values\n",
    "\n",
    "# Nonsense\n",
    "val_counts = trinuc_subs_oi[imp==3].index.value_counts()\n",
    "L.loc[ val_counts.index, 'Nonsense' ] = val_counts.values\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 515,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "8"
      ]
     },
     "execution_count": 515,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "int(26/3)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## write official function"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 564,
   "metadata": {},
   "outputs": [],
   "source": [
    "import logging\n",
    "logger = logging.getLogger(__name__)\n",
    "def get_all_possible_mutation_mat(seq, start = 1, end = None):\n",
    "    \"\"\"\n",
    "    seq: a string of nucleotides\n",
    "    start: the position of the coding exon sequence\n",
    "    \"\"\"\n",
    "    # by default, set end to the second to last position\n",
    "    if end is None:\n",
    "        end = len(seq) - 1\n",
    "    if start < 1: \n",
    "        raise ValueError('Start position must be >= 1.')\n",
    "    if end is None or end <= start:\n",
    "        raise ValueError('End position must be specified as greater than start position.')\n",
    "    if end >= len(seq):\n",
    "        raise ValueError('End position must be less than the length of the sequence.')\n",
    "    if (end-start+1) % 3 != 0 and 3 - ( (end-start+1) % 3 ) > (len(seq) - end): # if the last codon cannot be completed\n",
    "        raise ValueError('last codon cannot be completed with current setting. Please adjust the sequence and start/end positions.')\n",
    "    \n",
    "    seq = np.array([i for i in seq.upper()])\n",
    "    seq_1up = np.array( seq[start-1:end-1] ) # 1-bp upstream e.g. ATG\n",
    "    seq_mid = np.array( seq[start:end] ) # middle e.g. TGC\n",
    "    seq_1down = np.array( seq[start+1:end+1] ) # 1-bp downstream e.g. GCA\n",
    "    seq_length = len(seq_mid)\n",
    "\n",
    "    ##### ----- trinucleotide context ----- #####\n",
    "    indices = np.repeat([i for i in range(seq_length)], 3) # times 3 because for each single nucleotide, there are 3 ways to mutate it; e.g. array([0, 0, 0, 1, 1, 1, 2, 2, 2, 3, 3, 3, 4, 4, 4, 5, 5, 5])\n",
    "    old_trinuc = np.array(\n",
    "        paste(seq_1up[indices], seq_mid[indices], seq_1down[indices], sep = \"\") \n",
    "        ) # array(['CAT', 'CAT', 'CAT', 'ATG', 'ATG', 'ATG', 'TGT', 'TGT', 'TGT','GTC', 'GTC', 'GTC', 'TCC', 'TCC', 'TCC', 'CCA', 'CCA', 'CCA'])\n",
    "    new_bases = np.array(\n",
    "        [nts[nts != x].tolist() for x in seq_mid] \n",
    "        ).flatten() # generate all possible new bases at each position e.g. array(['C', 'G', 'T', 'A', 'C', 'G', 'A', 'C', 'T', 'A', 'C', 'G', 'A', 'G', 'T', 'A', 'G', 'T'])\n",
    "    new_trinuc = np.array(\n",
    "        paste(seq_1up[indices], new_bases, seq_1down[indices], sep=\"\") \n",
    "        ) # array(['CCT', 'CGT', 'CTT', 'AAG', 'ACG', 'AGG', 'TAT', 'TCT', 'TTT', 'GAC', 'GCC', 'GGC', 'TAC', 'TGC', 'TTC', 'CAA', 'CGA', 'CTA'])\n",
    "\n",
    "    ##### ----- codon ----- #####\n",
    "    num_codon = int(seq_length / 3) # get the lower integer\n",
    "    codon_start = np.repeat([i for i in range(0, num_codon * 3, 3)], 9) # 9 here because for each codon, there are 9 possible ways to mutate it\n",
    "    # e.g. array([0, 0, 0, 0, 0, 0, 0, 0, 0, 3, 3, 3, 3, 3, 3, 3, 3, 3])\n",
    "    pos_in_codon_to_mutate = np.tile(np.repeat([i for i in range(3)], 3), num_codon) # e.g. array([0, 0, 0, 1, 1, 1, 2, 2, 2, 0, 0, 0, 1, 1, 1, 2, 2, 2])\n",
    "    \n",
    "    if seq_length % 3 != 0:\n",
    "        logger.warning('Input coding sequence is not a multiple of 3. The last codon would be partial.')\n",
    "\n",
    "        codon_start = np.append( codon_start, [num_codon*3] * (seq_length % 3) * 3 )\n",
    "        pos_in_codon_to_mutate = np.append(pos_in_codon_to_mutate, np.repeat([i for i in range(seq_length % 3)], 3) ) # the last codon will only have 3 or 6 possible mutations\n",
    "\n",
    "    old_codon = np.array(\n",
    "        paste(seq[start + codon_start], seq[start + codon_start + 1], seq[start + codon_start + 2], sep=\"\")\n",
    "        ) # array(['ATG', 'ATG', 'ATG', 'ATG', 'ATG', 'ATG', 'ATG', 'ATG', 'ATG', 'TCC', 'TCC', 'TCC', 'TCC', 'TCC', 'TCC', 'TCC', 'TCC', 'TCC'])\n",
    "    \n",
    "\n",
    "    new_codon = np.array(\n",
    "        [__mutate(\n",
    "            old_codon[x], \n",
    "            pos_in_codon_to_mutate[x], \n",
    "            new_bases[x]\n",
    "            ) for x in range(len(pos_in_codon_to_mutate))]) # e.g. array(['CTG', 'GTG', 'TTG', 'AAG', 'ACG', 'AGG', 'ATA', 'ATC', 'ATT', 'ACC', 'CCC', 'GCC', 'TAC', 'TGC', 'TTC', 'TCA', 'TCG', 'TCT']\n",
    "\n",
    "    imp = np.array([ \n",
    "        IMPACT_MATRIX.loc[old_codon[x], new_codon[x]] \n",
    "        for x in range(len(old_codon)) \n",
    "        ]) # array([2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 1, 1, 1])\n",
    "\n",
    "    ##### ----- create L matrix ----- #####\n",
    "    trinuc_subs_oi = pd.DataFrame(\n",
    "        data = imp,\n",
    "        index = paste(old_trinuc, new_trinuc, sep=\">\"),\n",
    "        columns = ['IMPACT']\n",
    "        )\n",
    "\n",
    "    L = pd.DataFrame(index = trinuc_subs, columns = ['Synonymous', 'Missense', 'Nonsense'])\n",
    "    # Synonymous\n",
    "    val_counts = trinuc_subs_oi[imp==1].index.value_counts()\n",
    "    L.loc[ val_counts.index, 'Synonymous' ] = val_counts.values\n",
    "\n",
    "    # Missense\n",
    "    val_counts = trinuc_subs_oi[imp==2].index.value_counts()\n",
    "    L.loc[ val_counts.index, 'Missense' ] = val_counts.values\n",
    "\n",
    "    # Nonsense\n",
    "    val_counts = trinuc_subs_oi[imp==3].index.value_counts()\n",
    "    L.loc[ val_counts.index, 'Nonsense' ] = val_counts.values\n",
    "    return L"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### small test"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 565,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "WARNING:root:Input coding sequence is not a multiple of 3. The last codon would be partial.\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Synonymous</th>\n",
       "      <th>Missense</th>\n",
       "      <th>Nonsense</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>AAA&gt;ACA</th>\n",
       "      <td>NaN</td>\n",
       "      <td>5</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AAA&gt;AGA</th>\n",
       "      <td>1</td>\n",
       "      <td>4</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AAA&gt;ATA</th>\n",
       "      <td>NaN</td>\n",
       "      <td>3</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ACA&gt;AAA</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ACA&gt;AGA</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TGT&gt;TCT</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TGT&gt;TTT</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TTT&gt;TAT</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TTT&gt;TCT</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TTT&gt;TGT</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>192 rows × 3 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "        Synonymous Missense Nonsense\n",
       "AAA>ACA        NaN        5      NaN\n",
       "AAA>AGA          1        4      NaN\n",
       "AAA>ATA        NaN        3        2\n",
       "ACA>AAA        NaN      NaN      NaN\n",
       "ACA>AGA        NaN      NaN      NaN\n",
       "...            ...      ...      ...\n",
       "TGT>TCT        NaN      NaN      NaN\n",
       "TGT>TTT        NaN      NaN      NaN\n",
       "TTT>TAT        NaN      NaN      NaN\n",
       "TTT>TCT        NaN      NaN      NaN\n",
       "TTT>TGT        NaN      NaN      NaN\n",
       "\n",
       "[192 rows x 3 columns]"
      ]
     },
     "execution_count": 565,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "seq = 'AAAAAAAA'\n",
    "start = 1\n",
    "end = 6\n",
    "\n",
    "get_trans_mat(seq, start, end)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 544,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "WARNING:root:Input coding sequence is not a multiple of 3. The last codon would be partial.\n"
     ]
    }
   ],
   "source": [
    "seq_length = 8\n",
    "num_codon = int(seq_length / 3) # get the lower integer\n",
    "codon_start = np.repeat([i for i in range(0, num_codon * 3, 3)], 9) # 9 here because for each codon, there are 9 possible ways to mutate it\n",
    "# e.g. array([0, 0, 0, 0, 0, 0, 0, 0, 0, 3, 3, 3, 3, 3, 3, 3, 3, 3])\n",
    "pos_in_codon_to_mutate = np.tile(np.repeat([i for i in range(3)], 3), num_codon) # e.g. array([0, 0, 0, 1, 1, 1, 2, 2, 2, 0, 0, 0, 1, 1, 1, 2, 2, 2])\n",
    "\n",
    "if seq_length % 3 != 0:\n",
    "    logging.warning('Input coding sequence is not a multiple of 3. The last codon would be partial.')\n",
    "\n",
    "    codon_start = np.append( codon_start, [num_codon*3] * (seq_length % 3) * 3 )\n",
    "# codon_start\n",
    "    pos_in_codon_to_mutate = np.append(pos_in_codon_to_mutate, np.repeat([i for i in range(seq_length % 3)], 3) )\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# apply to panel DF"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [],
   "source": [
    "from tea.snv import get_all_possible_mutation_mat, TRINUC_SUBS, paste, __mutate, IMPACT_MATRIX\n",
    "import logging"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## test case"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### test the codon part of the function"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [
    {
     "ename": "AssertionError",
     "evalue": "",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mAssertionError\u001b[0m                            Traceback (most recent call last)",
      "\u001b[1;32m/Users/haochen/Desktop/Tapestri_analysis/panel_analysis/panel3359/dnds_analysis.ipynb Cell 33\u001b[0m in \u001b[0;36m<cell line: 14>\u001b[0;34m()\u001b[0m\n\u001b[1;32m     <a href='vscode-notebook-cell:/Users/haochen/Desktop/Tapestri_analysis/panel_analysis/panel3359/dnds_analysis.ipynb#X45sZmlsZQ%3D%3D?line=11'>12</a>\u001b[0m end_skip \u001b[39m=\u001b[39m (\u001b[39m3\u001b[39m\u001b[39m-\u001b[39morf_end\u001b[39m+\u001b[39moverlap_end) \u001b[39m%\u001b[39m \u001b[39m3\u001b[39m\n\u001b[1;32m     <a href='vscode-notebook-cell:/Users/haochen/Desktop/Tapestri_analysis/panel_analysis/panel3359/dnds_analysis.ipynb#X45sZmlsZQ%3D%3D?line=12'>13</a>\u001b[0m num_codon \u001b[39m=\u001b[39m (seq_length \u001b[39m-\u001b[39m start_skip \u001b[39m-\u001b[39m end_skip) \u001b[39m/\u001b[39m \u001b[39m3\u001b[39m\n\u001b[0;32m---> <a href='vscode-notebook-cell:/Users/haochen/Desktop/Tapestri_analysis/panel_analysis/panel3359/dnds_analysis.ipynb#X45sZmlsZQ%3D%3D?line=13'>14</a>\u001b[0m \u001b[39massert\u001b[39;00m(num_codon\u001b[39m.\u001b[39mis_integer())\n\u001b[1;32m     <a href='vscode-notebook-cell:/Users/haochen/Desktop/Tapestri_analysis/panel_analysis/panel3359/dnds_analysis.ipynb#X45sZmlsZQ%3D%3D?line=14'>15</a>\u001b[0m num_codon \u001b[39m=\u001b[39m \u001b[39mint\u001b[39m(num_codon)\n\u001b[1;32m     <a href='vscode-notebook-cell:/Users/haochen/Desktop/Tapestri_analysis/panel_analysis/panel3359/dnds_analysis.ipynb#X45sZmlsZQ%3D%3D?line=16'>17</a>\u001b[0m codon_start_ind \u001b[39m=\u001b[39m np\u001b[39m.\u001b[39mrepeat([i \u001b[39mfor\u001b[39;00m i \u001b[39min\u001b[39;00m \u001b[39mrange\u001b[39m(start_skip, seq_length\u001b[39m-\u001b[39mend_skip, \u001b[39m3\u001b[39m)], \u001b[39m9\u001b[39m) \u001b[39m# relative indices for codon start in the sequence; multiplies by 9 here because for each codon, there are 9 possible ways to mutate it\u001b[39;00m\n",
      "\u001b[0;31mAssertionError\u001b[0m: "
     ]
    }
   ],
   "source": [
    "test_str = amplicon_seq_df['AMPLICON-seq'][3]\n",
    "seq = np.array([i for i in test_str])\n",
    "\n",
    "orf_start=23\n",
    "overlap_start=23\n",
    "overlap_end=245\n",
    "orf_end=246\n",
    "\n",
    "seq_length = overlap_end - overlap_start + 1\n",
    "\n",
    "start_skip = (3-overlap_start+orf_start) % 3\n",
    "end_skip = (3-orf_end+overlap_end) % 3\n",
    "num_codon = (seq_length - start_skip - end_skip) / 3\n",
    "assert(num_codon.is_integer())\n",
    "num_codon = int(num_codon)\n",
    "\n",
    "codon_start_ind = np.repeat([i for i in range(start_skip, seq_length-end_skip, 3)], 9) # relative indices for codon start in the sequence; multiplies by 9 here because for each codon, there are 9 possible ways to mutate it\n",
    "# e.g. array([0, 0, 0, 0, 0, 0, 0, 0, 0, 3, 3, 3, 3, 3, 3, 3, 3, 3])\n",
    "pos_in_codon_to_mutate = np.tile(np.repeat([i for i in range(3)], 3), num_codon) # e.g. array([0, 0, 0, 1, 1, 1, 2, 2, 2, 0, 0, 0, 1, 1, 1, 2, 2, 2])\n",
    "\n",
    "if start_skip > 0 or end_skip > 0:\n",
    "    # logger.debug('the first/last codon would be partially considered for mutation.')\n",
    "\n",
    "    codon_start_ind = np.concatenate([\n",
    "        [start_skip - 3] * (start_skip) * 3, \n",
    "        codon_start_ind, \n",
    "        [start_skip + num_codon*3] * (end_skip) * 3\n",
    "        ]).astype(int)\n",
    "    pos_in_codon_to_mutate = np.concatenate([\n",
    "        np.repeat([i for i in range(3-start_skip, 3)], 3),\n",
    "        pos_in_codon_to_mutate, \n",
    "        np.repeat([i for i in range(end_skip)], 3) \n",
    "        ]).astype(int) # the last codon will only have 3 or 6 possible mutations\n",
    "\n",
    "old_codon = np.array(\n",
    "    paste(seq[overlap_start + codon_start_ind], seq[overlap_start + codon_start_ind + 1], seq[overlap_start + codon_start_ind + 2], sep=\"\")\n",
    "    ) # array(['ATG', 'ATG', 'ATG', 'ATG', 'ATG', 'ATG', 'ATG', 'ATG', 'ATG', 'TCC', 'TCC', 'TCC', 'TCC', 'TCC', 'TCC', 'TCC', 'TCC', 'TCC'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "2"
      ]
     },
     "execution_count": 18,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "end_skip"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 57,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(354,)"
      ]
     },
     "execution_count": 57,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "pos_in_codon_to_mutate.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 58,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(118,)"
      ]
     },
     "execution_count": 58,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "seq_mid.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 59,
   "metadata": {},
   "outputs": [],
   "source": [
    "nts = np.array(['A', 'C', 'G', 'T'])\n",
    "seq_mid = np.array( seq[overlap_start:overlap_end+1] )\n",
    "new_bases = np.array(\n",
    "    [nts[nts != x].tolist() for x in seq_mid] \n",
    "    ).flatten() # generate all possible new bases at each position e.g. array(['C', 'G', 'T', 'A', 'C', 'G', 'A', 'C', 'T', 'A', 'C', 'G', 'A', 'G', 'T', 'A', 'G', 'T'])\n",
    "\n",
    "new_codon = np.array(\n",
    "    [__mutate(\n",
    "    old_codon[x], \n",
    "    pos_in_codon_to_mutate[x], \n",
    "    new_bases[x]\n",
    "    ) for x in range(len(pos_in_codon_to_mutate))\n",
    "    ]) # e.g. array(['CTG', 'GTG', 'TTG', 'AAG', 'ACG', 'AGG', 'ATA', 'ATC', 'ATT', 'ACC', 'CCC', 'GCC', 'TAC', 'TGC', 'TTC', 'TCA', 'TCG', 'TCT']\n",
    "\n",
    "imp = np.array([ \n",
    "    IMPACT_MATRIX.loc[old_codon[x], new_codon[x]] \n",
    "    for x in range(len(old_codon)) \n",
    "    ]) # array([2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 1, 1, 1])"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### test whole function"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [],
   "source": [
    "test_str = 'ATGCATGCATGC'\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [],
   "source": [
    "from tea.snv import __mutate, IMPACT_MATRIX\n",
    "nts = np.array(['A', 'T', 'G', 'C'])\n",
    "L = get_all_possible_mutation_mat(test_str, 3, 4, 11, 10)\n",
    "# L.to_csv('test.csv')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Synonymous</th>\n",
       "      <th>Missense</th>\n",
       "      <th>Nonsense</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>AAA&gt;ACA</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AAA&gt;AGA</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AAA&gt;ATA</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ACA&gt;AAA</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ACA&gt;AGA</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TGT&gt;TCT</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TGT&gt;TTT</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TTT&gt;TAT</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TTT&gt;TCT</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TTT&gt;TGT</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>192 rows × 3 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "        Synonymous Missense Nonsense\n",
       "AAA>ACA        NaN      NaN      NaN\n",
       "AAA>AGA        NaN      NaN      NaN\n",
       "AAA>ATA        NaN      NaN      NaN\n",
       "ACA>AAA        NaN      NaN      NaN\n",
       "ACA>AGA        NaN      NaN      NaN\n",
       "...            ...      ...      ...\n",
       "TGT>TCT        NaN      NaN      NaN\n",
       "TGT>TTT        NaN      NaN      NaN\n",
       "TTT>TAT        NaN      NaN      NaN\n",
       "TTT>TCT        NaN      NaN      NaN\n",
       "TTT>TGT        NaN      NaN      NaN\n",
       "\n",
       "[192 rows x 3 columns]"
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "L"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## actaual thing"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 139,
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_all_possible_mutation_mat(seq, orf_start, overlap_start, orf_end, overlap_end, strand = '+', log_level = logging.INFO):\n",
    "    \"\"\"\n",
    "    seq: a string of nucleotides\n",
    "    (all the positions need to be 0 based)\n",
    "    orf_start: the position of the coding exon sequence\n",
    "    overlap_start: the position of the overlap sequence\n",
    "    orf_end: the position of the coding exon sequence\n",
    "    overlap_end: the position of the overlap sequence\n",
    "    strand: the strand of the sequence\n",
    "\n",
    "    \"\"\"\n",
    "    logger = logging.getLogger(__name__)\n",
    "    logger.setLevel(log_level)\n",
    "\n",
    "    # by default, set orf_end to the second to last position\n",
    "    if orf_start < 1: \n",
    "        raise ValueError('ORF start position must be >= 1.')\n",
    "    if overlap_start < orf_start:\n",
    "        raise ValueError('Overlap start must be >= ORF start.')\n",
    "    if orf_end <= overlap_start:\n",
    "        raise ValueError('ORF end must be > overlap start.')\n",
    "    if overlap_end > orf_end:\n",
    "        raise ValueError('Overlap end must be <= ORF end.')\n",
    "    if orf_end > len(seq):\n",
    "        raise ValueError('End position must be less than the length of the sequence.')\n",
    "    # if (orf_end-orf_start+1) % 3 != 0 and 3 - ( (orf_end-orf_start+1) % 3 ) > (len(seq) - orf_end): # if the last codon cannot be completed\n",
    "    #     raise ValueError('last codon cannot be completed with current setting. Please adjust the sequence and orf_start/orf_end positions.')\n",
    "    \n",
    "    seq = np.array([i for i in seq.upper()])\n",
    "    if strand == '-':\n",
    "        # reverse strand\n",
    "        seq = seq[::-1]\n",
    "    seq_1up = np.array( seq[overlap_start-1:overlap_end] ) # 1-bp upstream e.g. ATG\n",
    "    seq_mid = np.array( seq[overlap_start:overlap_end+1] ) # middle e.g. TGC\n",
    "    seq_1down = np.array( seq[overlap_start+1:overlap_end+2] ) # 1-bp downstream e.g. GCA\n",
    "    seq_length = len(seq_mid)\n",
    "\n",
    "    ##### ----- trinucleotide context ----- #####\n",
    "    indices = np.repeat([i for i in range(seq_length)], 3) # times 3 because for each single nucleotide, there are 3 ways to mutate it; e.g. array([0, 0, 0, 1, 1, 1, 2, 2, 2, 3, 3, 3, 4, 4, 4, 5, 5, 5])\n",
    "    old_trinuc = np.array(\n",
    "        paste(seq_1up[indices], seq_mid[indices], seq_1down[indices], sep = \"\") \n",
    "        ) # array(['CAT', 'CAT', 'CAT', 'ATG', 'ATG', 'ATG', 'TGT', 'TGT', 'TGT','GTC', 'GTC', 'GTC', 'TCC', 'TCC', 'TCC', 'CCA', 'CCA', 'CCA'])\n",
    "    new_bases = np.array(\n",
    "        [nts[nts != x].tolist() for x in seq_mid] \n",
    "        ).flatten() # generate all possible new bases at each position e.g. array(['C', 'G', 'T', 'A', 'C', 'G', 'A', 'C', 'T', 'A', 'C', 'G', 'A', 'G', 'T', 'A', 'G', 'T'])\n",
    "    new_trinuc = np.array(\n",
    "        paste(seq_1up[indices], new_bases, seq_1down[indices], sep=\"\") \n",
    "        ) # array(['CCT', 'CGT', 'CTT', 'AAG', 'ACG', 'AGG', 'TAT', 'TCT', 'TTT', 'GAC', 'GCC', 'GGC', 'TAC', 'TGC', 'TTC', 'CAA', 'CGA', 'CTA'])\n",
    "\n",
    "    ##### ----- codon ----- #####\n",
    "    start_skip = (3-overlap_start+orf_start) % 3 # number of bps to skip at the beginning\n",
    "    end_skip = (3-orf_end+overlap_end) % 3 # number of bps to skip at the end\n",
    "    num_codon = (seq_length - start_skip - end_skip) / 3\n",
    "    assert num_codon.is_integer(), f'the input ORF is not a multiple of 3; seq_length {seq_length}, start_skip {start_skip}, end_skip {end_skip}, num_codon {num_codon}'\n",
    "    num_codon = int(num_codon)\n",
    "\n",
    "    codon_start_ind = np.repeat([i for i in range(start_skip, seq_length-end_skip, 3)], 9) # relative indices for codon start in the sequence; multiplies by 9 here because for each codon, there are 9 possible ways to mutate it\n",
    "    # e.g. array([0, 0, 0, 0, 0, 0, 0, 0, 0, 3, 3, 3, 3, 3, 3, 3, 3, 3])\n",
    "    pos_in_codon_to_mutate = np.tile(np.repeat([i for i in range(3)], 3), num_codon) # e.g. array([0, 0, 0, 1, 1, 1, 2, 2, 2, 0, 0, 0, 1, 1, 1, 2, 2, 2])\n",
    "\n",
    "    if start_skip > 0 or end_skip > 0:\n",
    "        logger.debug('the first/last codon would be partially considered for mutation.')\n",
    "\n",
    "        codon_start_ind = np.concatenate([\n",
    "            [start_skip - 3] * (start_skip) * 3, \n",
    "            codon_start_ind, \n",
    "            [start_skip + num_codon*3] * (end_skip) * 3\n",
    "            ]).astype(int)\n",
    "        pos_in_codon_to_mutate = np.concatenate([\n",
    "            np.repeat([i for i in range(start_skip)], 3),\n",
    "            pos_in_codon_to_mutate, \n",
    "            np.repeat([i for i in range(end_skip)], 3) \n",
    "            ]).astype(int) # the last codon will only have 3 or 6 possible mutations\n",
    "\n",
    "    old_codon = np.array(\n",
    "        paste(seq[overlap_start + codon_start_ind], seq[overlap_start + codon_start_ind + 1], seq[overlap_start + codon_start_ind + 2], sep=\"\")\n",
    "        ) # array(['ATG', 'ATG', 'ATG', 'ATG', 'ATG', 'ATG', 'ATG', 'ATG', 'ATG', 'TCC', 'TCC', 'TCC', 'TCC', 'TCC', 'TCC', 'TCC', 'TCC', 'TCC'])\n",
    "    \n",
    "    new_codon = np.array(\n",
    "        [__mutate(\n",
    "            old_codon[x], \n",
    "            pos_in_codon_to_mutate[x], \n",
    "            new_bases[x]\n",
    "            ) for x in range(len(pos_in_codon_to_mutate))]) # e.g. array(['CTG', 'GTG', 'TTG', 'AAG', 'ACG', 'AGG', 'ATA', 'ATC', 'ATT', 'ACC', 'CCC', 'GCC', 'TAC', 'TGC', 'TTC', 'TCA', 'TCG', 'TCT']\n",
    "\n",
    "    imp = np.array([ \n",
    "        IMPACT_MATRIX.loc[old_codon[x], new_codon[x]] \n",
    "        for x in range(len(old_codon)) \n",
    "        ]) # array([2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 1, 1, 1])\n",
    "\n",
    "    ##### ----- create L matrix ----- #####\n",
    "    trinuc_subs_oi = pd.DataFrame(\n",
    "        data = imp,\n",
    "        index = paste(old_trinuc, new_trinuc, sep=\">\"),\n",
    "        columns = ['IMPACT']\n",
    "        )\n",
    "\n",
    "    L = pd.DataFrame(index = TRINUC_SUBS, columns = ['Synonymous', 'Missense', 'Nonsense'])\n",
    "    # Synonymous\n",
    "    val_counts = trinuc_subs_oi[imp==1].index.value_counts()\n",
    "    L.loc[ val_counts.index, 'Synonymous' ] = val_counts.values\n",
    "\n",
    "    # Missense\n",
    "    val_counts = trinuc_subs_oi[imp==2].index.value_counts()\n",
    "    L.loc[ val_counts.index, 'Missense' ] = val_counts.values\n",
    "\n",
    "    # Nonsense\n",
    "    val_counts = trinuc_subs_oi[imp==3].index.value_counts()\n",
    "    L.loc[ val_counts.index, 'Nonsense' ] = val_counts.values\n",
    "    return L"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 144,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[WARNING] L mat calculation encountered ValueError for RPL5 AMPL250226\n",
      "[WARNING] L mat calculation encountered ValueError for CDKN2A AMPL113204\n",
      "[WARNING] L mat calculation encountered ValueError for CDKN2A AMPL113204\n"
     ]
    }
   ],
   "source": [
    "# from tea.snv import get_all_possible_mutation_mat, TRINUC_SUBS, paste, __mutate, IMPACT_MATRIX\n",
    "\n",
    "logger = logging.getLogger(__name__)\n",
    "\n",
    "gene_L_mat_map = {}\n",
    "cds_l = []\n",
    "# iterate through every unique gene\n",
    "for gene_i in panel_df['CDS-gene_name'].unique():\n",
    "    # ##########################\n",
    "    # # @DEBUG\n",
    "    # if not gene_i == 'KRAS':\n",
    "    #     continue\n",
    "    # ########################## \n",
    "    gene_df = panel_df[panel_df['CDS-gene_name'] == gene_i]\n",
    "    gene_L_mat = pd.DataFrame(index = TRINUC_SUBS, columns = ['Synonymous', 'Missense', 'Nonsense'])\n",
    "\n",
    "    # iterate through every unique CDS of this gene, covered by the panel\n",
    "    for _, row in gene_df.iterrows():\n",
    "        amplicon_id = row['AMPLICON-ID']\n",
    "\n",
    "        # ----- exon information -----\n",
    "        # need to covert all the coordinates to 0-based\n",
    "        exon_unique_id = row['CDS-exon_unique_id']\n",
    "        if exon_unique_id == '.-.':\n",
    "            logger.debug(f'{gene_i} {amplicon_id} has no exon information; skipping.')\n",
    "            continue\n",
    "        exon_strand = row['CDS-UCSC_strand']\n",
    "        cds_start = row['CDS-start']\n",
    "        cds_end = row['CDS-end'] - 1\n",
    "        overlap_start = row['overlap_start'] \n",
    "        overlap_end = row['overlap_end'] - 1\n",
    "        \n",
    "        # ----- get coding frame -----\n",
    "        # remove incomplete codons, so that the CDS is in frame\n",
    "        start_r_bp = row['start_r_bp']\n",
    "        end_r_bp = row['end_r_bp']\n",
    "        if exon_strand == '+':\n",
    "            cds_start += start_r_bp\n",
    "            cds_end -= end_r_bp\n",
    "        else:\n",
    "            cds_start += end_r_bp\n",
    "            cds_end -= start_r_bp\n",
    "        assert (cds_end - cds_start + 1) % 3 == 0, 'CDS sequence not in frame!'\n",
    "\n",
    "        if overlap_start < cds_start:\n",
    "            # logger.debug('overlap_start < cds_start')\n",
    "            logger.debug(f'for overlap_start < cds_start for {gene_i} {amplicon_id}')\n",
    "            logger.debug(f'overlap_start: {overlap_start}, cds_start: {cds_start}')\n",
    "            # raise ValueError\n",
    "            overlap_start = cds_start\n",
    "        if overlap_end > cds_end:\n",
    "            logger.debug(f'[WARNING] overlap_end > cds_end for {gene_i} {amplicon_id}')\n",
    "            logger.debug(f'overlap_end: {overlap_end}, cds_end: {cds_end}')\n",
    "            overlap_end = cds_end\n",
    "        # move to the codon closest to the overlap\n",
    "        covered_cds_start = overlap_start - (overlap_start - cds_start) % 3\n",
    "        covered_cds_end = overlap_end + (cds_end - overlap_end) % 3\n",
    "        \n",
    "        # ----- covering amplicon information -----\n",
    "        amplicon_record = amplicon_seq_df[amplicon_seq_df['AMPLICON-ID'] == amplicon_id]\n",
    "        amplicon_seq = amplicon_record['AMPLICON-seq'].iloc[0]\n",
    "        amplicon_start = amplicon_record['AMPLICON-start'].iloc[0]\n",
    "        amplicon_end = amplicon_record['AMPLICON-end'].iloc[0]\n",
    "\n",
    "        # relative position in the amplicon sequence \n",
    "        overlap_relative_start_in_amp = int( overlap_start - amplicon_start )\n",
    "        overlap_relative_end_in_amp = int( overlap_end - amplicon_start )\n",
    "        orf_relative_start_in_amp = int( covered_cds_start - amplicon_start )\n",
    "        orf_relative_end_in_amp = int( covered_cds_end - amplicon_start )\n",
    "        # overlap_seq = amplicon_seq[overlap_seq_start:overlap_seq_end+1]\n",
    "        try:\n",
    "            L = get_all_possible_mutation_mat(\n",
    "                amplicon_seq, \n",
    "                orf_start = orf_relative_start_in_amp, \n",
    "                overlap_start = overlap_relative_start_in_amp, \n",
    "                overlap_end = overlap_relative_end_in_amp,\n",
    "                orf_end = orf_relative_end_in_amp,\n",
    "                strand = strand\n",
    "                )\n",
    "            # print(L.sum(axis=0))\n",
    "        except ValueError:\n",
    "            print(f'[WARNING] L mat calculation encountered ValueError for {gene_i} {amplicon_id}')\n",
    "            continue\n",
    "            \n",
    "        gene_L_mat = gene_L_mat.add(L, fill_value=0)\n",
    "        # print(gene_L_mat.sum(axis=0))\n",
    "    gene_L_mat_map[gene_i] = gene_L_mat"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# sanity checks"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 118,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>INSERT-chr</th>\n",
       "      <th>INSERT-start</th>\n",
       "      <th>INSERT-end</th>\n",
       "      <th>AMPLICON-ID</th>\n",
       "      <th>CDS-chr</th>\n",
       "      <th>CDS-start</th>\n",
       "      <th>CDS-end</th>\n",
       "      <th>CDS-Ensembl_ID</th>\n",
       "      <th>CDS-UCSC_ID</th>\n",
       "      <th>CDS-UCSC_index</th>\n",
       "      <th>CDS-UCSC_strand</th>\n",
       "      <th>CDS-gene_name</th>\n",
       "      <th>bp_overlap</th>\n",
       "      <th>overlap_start</th>\n",
       "      <th>overlap_end</th>\n",
       "      <th>CDS-exon_unique_id</th>\n",
       "      <th>CDS-ORF-frame</th>\n",
       "      <th>start_r_bp</th>\n",
       "      <th>end_r_bp</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>6252892</td>\n",
       "      <td>6253107</td>\n",
       "      <td>AMPL250217</td>\n",
       "      <td>chr1</td>\n",
       "      <td>6252989</td>\n",
       "      <td>6253114</td>\n",
       "      <td>ENST00000234875</td>\n",
       "      <td>uc001amd.3</td>\n",
       "      <td>1</td>\n",
       "      <td>-</td>\n",
       "      <td>RPL22</td>\n",
       "      <td>118</td>\n",
       "      <td>6252989</td>\n",
       "      <td>6253107</td>\n",
       "      <td>uc001amd.3-1</td>\n",
       "      <td>2.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>2.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1</td>\n",
       "      <td>6257606</td>\n",
       "      <td>6257826</td>\n",
       "      <td>AMPL257565</td>\n",
       "      <td>chr1</td>\n",
       "      <td>6257711</td>\n",
       "      <td>6257816</td>\n",
       "      <td>ENST00000234875</td>\n",
       "      <td>uc001amd.3</td>\n",
       "      <td>2</td>\n",
       "      <td>-</td>\n",
       "      <td>RPL22</td>\n",
       "      <td>105</td>\n",
       "      <td>6257711</td>\n",
       "      <td>6257816</td>\n",
       "      <td>uc001amd.3-2</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr1</td>\n",
       "      <td>11204888</td>\n",
       "      <td>11205104</td>\n",
       "      <td>AMPL257566</td>\n",
       "      <td>chr1</td>\n",
       "      <td>11205024</td>\n",
       "      <td>11205102</td>\n",
       "      <td>ENST00000361445</td>\n",
       "      <td>uc001asd.3</td>\n",
       "      <td>25</td>\n",
       "      <td>-</td>\n",
       "      <td>MTOR</td>\n",
       "      <td>78</td>\n",
       "      <td>11205024</td>\n",
       "      <td>11205102</td>\n",
       "      <td>uc001asd.3-25</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr1</td>\n",
       "      <td>11288734</td>\n",
       "      <td>11288957</td>\n",
       "      <td>AMPL253201</td>\n",
       "      <td>chr1</td>\n",
       "      <td>11288724</td>\n",
       "      <td>11288975</td>\n",
       "      <td>ENST00000361445</td>\n",
       "      <td>uc001asd.3</td>\n",
       "      <td>39</td>\n",
       "      <td>-</td>\n",
       "      <td>MTOR</td>\n",
       "      <td>223</td>\n",
       "      <td>11288734</td>\n",
       "      <td>11288957</td>\n",
       "      <td>uc001asd.3-39</td>\n",
       "      <td>2.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>2.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr1</td>\n",
       "      <td>11297897</td>\n",
       "      <td>11298107</td>\n",
       "      <td>AMPL257567</td>\n",
       "      <td>chr1</td>\n",
       "      <td>11297899</td>\n",
       "      <td>11298105</td>\n",
       "      <td>ENST00000361445</td>\n",
       "      <td>uc001asd.3</td>\n",
       "      <td>45</td>\n",
       "      <td>-</td>\n",
       "      <td>MTOR</td>\n",
       "      <td>206</td>\n",
       "      <td>11297899</td>\n",
       "      <td>11298105</td>\n",
       "      <td>uc001asd.3-45</td>\n",
       "      <td>2.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>2.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>630</th>\n",
       "      <td>chrX</td>\n",
       "      <td>128605116</td>\n",
       "      <td>128605306</td>\n",
       "      <td>AMPL257734</td>\n",
       "      <td>chrX</td>\n",
       "      <td>128605180</td>\n",
       "      <td>128605303</td>\n",
       "      <td>ENST00000371122</td>\n",
       "      <td>uc004eun.4</td>\n",
       "      <td>5</td>\n",
       "      <td>-</td>\n",
       "      <td>SMARCA1</td>\n",
       "      <td>123</td>\n",
       "      <td>128605180</td>\n",
       "      <td>128605303</td>\n",
       "      <td>uc004eun.4-5</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>631</th>\n",
       "      <td>chrX</td>\n",
       "      <td>135957311</td>\n",
       "      <td>135957532</td>\n",
       "      <td>AMPL250434</td>\n",
       "      <td>chrX</td>\n",
       "      <td>135957244</td>\n",
       "      <td>135957327</td>\n",
       "      <td>ENST00000320676</td>\n",
       "      <td>uc004fae.2</td>\n",
       "      <td>1</td>\n",
       "      <td>-</td>\n",
       "      <td>RBMX</td>\n",
       "      <td>16</td>\n",
       "      <td>135957311</td>\n",
       "      <td>135957327</td>\n",
       "      <td>uc004fae.2-1</td>\n",
       "      <td>2.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>2.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>632</th>\n",
       "      <td>chrX</td>\n",
       "      <td>135957311</td>\n",
       "      <td>135957532</td>\n",
       "      <td>AMPL250434</td>\n",
       "      <td>chrX</td>\n",
       "      <td>135957416</td>\n",
       "      <td>135957542</td>\n",
       "      <td>ENST00000320676</td>\n",
       "      <td>uc004fae.2</td>\n",
       "      <td>2</td>\n",
       "      <td>-</td>\n",
       "      <td>RBMX</td>\n",
       "      <td>116</td>\n",
       "      <td>135957416</td>\n",
       "      <td>135957532</td>\n",
       "      <td>uc004fae.2-2</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>633</th>\n",
       "      <td>chrX</td>\n",
       "      <td>153581295</td>\n",
       "      <td>153581504</td>\n",
       "      <td>AMPL250435</td>\n",
       "      <td>chrX</td>\n",
       "      <td>153581368</td>\n",
       "      <td>153581572</td>\n",
       "      <td>ENST00000422373</td>\n",
       "      <td>uc004fkk.2</td>\n",
       "      <td>10</td>\n",
       "      <td>-</td>\n",
       "      <td>FLNA</td>\n",
       "      <td>136</td>\n",
       "      <td>153581368</td>\n",
       "      <td>153581504</td>\n",
       "      <td>uc004fkk.2-10</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>634</th>\n",
       "      <td>chrX</td>\n",
       "      <td>154299804</td>\n",
       "      <td>154299956</td>\n",
       "      <td>AMPL251376</td>\n",
       "      <td>chrX</td>\n",
       "      <td>154299802</td>\n",
       "      <td>154299925</td>\n",
       "      <td>ENST00000369462</td>\n",
       "      <td>uc004fna.3</td>\n",
       "      <td>0</td>\n",
       "      <td>+</td>\n",
       "      <td>BRCC3</td>\n",
       "      <td>121</td>\n",
       "      <td>154299804</td>\n",
       "      <td>154299925</td>\n",
       "      <td>uc004fna.3-0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>635 rows × 19 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "    INSERT-chr  INSERT-start  INSERT-end AMPLICON-ID CDS-chr  CDS-start  \\\n",
       "0         chr1       6252892     6253107  AMPL250217    chr1    6252989   \n",
       "1         chr1       6257606     6257826  AMPL257565    chr1    6257711   \n",
       "2         chr1      11204888    11205104  AMPL257566    chr1   11205024   \n",
       "3         chr1      11288734    11288957  AMPL253201    chr1   11288724   \n",
       "4         chr1      11297897    11298107  AMPL257567    chr1   11297899   \n",
       "..         ...           ...         ...         ...     ...        ...   \n",
       "630       chrX     128605116   128605306  AMPL257734    chrX  128605180   \n",
       "631       chrX     135957311   135957532  AMPL250434    chrX  135957244   \n",
       "632       chrX     135957311   135957532  AMPL250434    chrX  135957416   \n",
       "633       chrX     153581295   153581504  AMPL250435    chrX  153581368   \n",
       "634       chrX     154299804   154299956  AMPL251376    chrX  154299802   \n",
       "\n",
       "       CDS-end   CDS-Ensembl_ID CDS-UCSC_ID CDS-UCSC_index CDS-UCSC_strand  \\\n",
       "0      6253114  ENST00000234875  uc001amd.3              1               -   \n",
       "1      6257816  ENST00000234875  uc001amd.3              2               -   \n",
       "2     11205102  ENST00000361445  uc001asd.3             25               -   \n",
       "3     11288975  ENST00000361445  uc001asd.3             39               -   \n",
       "4     11298105  ENST00000361445  uc001asd.3             45               -   \n",
       "..         ...              ...         ...            ...             ...   \n",
       "630  128605303  ENST00000371122  uc004eun.4              5               -   \n",
       "631  135957327  ENST00000320676  uc004fae.2              1               -   \n",
       "632  135957542  ENST00000320676  uc004fae.2              2               -   \n",
       "633  153581572  ENST00000422373  uc004fkk.2             10               -   \n",
       "634  154299925  ENST00000369462  uc004fna.3              0               +   \n",
       "\n",
       "    CDS-gene_name  bp_overlap  overlap_start  overlap_end CDS-exon_unique_id  \\\n",
       "0           RPL22         118        6252989      6253107       uc001amd.3-1   \n",
       "1           RPL22         105        6257711      6257816       uc001amd.3-2   \n",
       "2            MTOR          78       11205024     11205102      uc001asd.3-25   \n",
       "3            MTOR         223       11288734     11288957      uc001asd.3-39   \n",
       "4            MTOR         206       11297899     11298105      uc001asd.3-45   \n",
       "..            ...         ...            ...          ...                ...   \n",
       "630       SMARCA1         123      128605180    128605303       uc004eun.4-5   \n",
       "631          RBMX          16      135957311    135957327       uc004fae.2-1   \n",
       "632          RBMX         116      135957416    135957532       uc004fae.2-2   \n",
       "633          FLNA         136      153581368    153581504      uc004fkk.2-10   \n",
       "634         BRCC3         121      154299804    154299925       uc004fna.3-0   \n",
       "\n",
       "    CDS-ORF-frame  start_r_bp  end_r_bp  \n",
       "0             2.0         0.0       2.0  \n",
       "1             0.0         0.0       0.0  \n",
       "2             0.0         0.0       0.0  \n",
       "3             2.0         0.0       2.0  \n",
       "4             2.0         0.0       2.0  \n",
       "..            ...         ...       ...  \n",
       "630           0.0         0.0       0.0  \n",
       "631           2.0         0.0       2.0  \n",
       "632           0.0         0.0       0.0  \n",
       "633           0.0         0.0       0.0  \n",
       "634           0.0         0.0       0.0  \n",
       "\n",
       "[635 rows x 19 columns]"
      ]
     },
     "execution_count": 118,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "panel_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 119,
   "metadata": {},
   "outputs": [],
   "source": [
    "for gene_i in panel_df['CDS-gene_name'].unique():\n",
    "    if gene_i not in gene_L_mat_map:\n",
    "        print(f'[WARNING] {gene_i} not in gene_L_mat_map')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 146,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Synonymous     373\n",
       "Missense      1231\n",
       "Nonsense        66\n",
       "dtype: object"
      ]
     },
     "execution_count": 146,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "gene_L_mat_map['KRAS'].sum(axis=0)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Pickle the gene L matrices"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 147,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pickle\n",
    "from pathlib import Path\n",
    "output_dir = Path('/Users/haochen/Desktop/Tapestri_analysis/dnds')\n",
    "with open(output_dir / 'panel3359_gene_L_mat_map.pkl', 'wb') as f:\n",
    "    pickle.dump(gene_L_mat_map, f)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 148,
   "metadata": {},
   "outputs": [],
   "source": [
    "with open(output_dir / 'panel3359_gene_L_mat_map.pkl', 'rb') as f:\n",
    "    gene_L_mat_map = pickle.load(f)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 149,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Synonymous     373\n",
       "Missense      1231\n",
       "Nonsense        66\n",
       "dtype: object"
      ]
     },
     "execution_count": 149,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "gene_L_mat_map['KRAS'].sum(axis=0)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3.8.13 ('mosaic-custom')",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.13"
  },
  "orig_nbformat": 4,
  "vscode": {
   "interpreter": {
    "hash": "3c6974fee5b6fa730f6923551aa02af44f42e2bd0ff1a6675c7e323f195d194a"
   }
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
